### ISSUE TITLE
As service author, I can specify in a manifest the strategy for additional parameters in the backing service

### ISSUE THREAD

**Issue by pivotal-dylan-roberts** [spring-cloud-incubator/spring-cloud-app-broker#139]
`AS` service author, I can specify in a manifest the strategy for additional parameters in the backing service that gets deployed to CF on service create
`SO THAT` I can customize the backing services parameters dynamically.

`GIVEN` the application manifest:
```
spring:
  cloud
    appbroker:
      services: 
        apps:
        - name: app-with-services
           services:
           - service-instance-name: my-db-service
            parameters-transformers:
             - name: ParameterMapping
               include: paramA,paramC
        services:
        - service-instance-name: my-db-service
          name: mysqldb
          plan: small
```
`WHEN` a service gets created with `cf create-service my-service -c '{"paramA":"valueA", "paramB":"valueB", "paramC":"valueC"}'`
`THEN` the SI `my-db-service` will be created with the parameters `-c '{"paramA":"valueA", "paramC":"valueC"}'`

**Comment by scottfrederick** [comment]
Depends on #14

**Comment by scottfrederick** [comment]
@pivotal-dylan-roberts Does your use case require that the parameters the user provides in the create-service parameters be transformed in any way before being provided to the backing service instance (e.g. changing keys provided in the JSON hash), or can the parameter keys and values be copied as-is?

**Comment by AlbertoImpl** [comment]
Will this match your expection @pivotal-dylan-roberts ? 

`GIVEN`
```
spring:
  cloud
    appbroker:
      services: 
        apps:
        - name: app-with-services
           services:
           - service-instance-name: my-db-service
             service-parameters-transformers: InputApp1Transformer
        services:
        - service-instance-name: my-db-service
          name: mysqldb
          plan: small
```
`AND` `InputApp1Transformer` is specified to transform a value A into a value B
`WHEN` `cf create-service my-service -c '{"paramA":"valueA"}'`
`THEN` the SI `my-db-service` will be created with `-c '{"paramA":"valueB"}'`


Feel free to update it, specially the `THEN` clause since we can't really validate it.

Also, as Scott mentioned, would this suffice as a first story?
`GIVEN`
```
spring:
  cloud
    appbroker:
      services: 
        apps:
        - name: app-with-services
           services:
           - service-instance-name: my-db-service
        services:
        - service-instance-name: my-db-service
          name: mysqldb
          plan: small
```
`WHEN` `cf create-service my-service -c '{"paramA":"valueA"}'`
`THEN` the SI `my-db-service` will be created with `-c '{"paramA":"valueA"}'`

**Comment by pivotal-dylan-roberts** [comment]
@AlbertoImpl @scottfrederick @LittleBaiBai The last option (ie. just a straight pass through of all `-c` params) will get us unblocked for SCS3; so that's all we are requesting right now!

**Comment by pivotal-dylan-roberts** [comment]
@AlbertoImpl that would be perfect. Leaves us some flexibility, just in case.

**Comment by scottfrederick** [comment]
@AlbertoImpl I'm OK with this as a variation of the existing `ParametersTransformer` pattern. I do think that all of these parameter transformers should have at the least a configurable white-list of parameters that they will transform (i.e. copy to the backing SI params in this case). Without that minimal filtering, anything the user puts into the `cf create-service -c '{}'` params can end up in the backing SI params.

**Comment by AlbertoImpl** [comment]
Updated the story to do both, adding a default strategy to pass all the params @pivotal-dylan-roberts , but with the filtering @scottfrederick  suggested. Let me know if you think it can be improved or if it won't suffice.
cc: @LittleBaiBai

**Comment by pivotal-dylan-roberts** [comment]
@AlbertoImpl Is there an option to skip the filter altogether? Maybe just leaving out the `include` property? Otherwise we'll have to manually whitelist all of the config-server configuration properties. 

Also, would nested properties be handled by the `include`?

cc: @LittleBaiBai

**Comment by AlbertoImpl** [comment]
I still believe that we should be whitelisting only so that other parameters do no get passed through.
You could add only the object roots since they are limited in number.

**Comment by scottfrederick** [comment]
@pivotal-dylan-roberts The whitelisting should be mandatory (i.e. an empty `include` parameter will result in no parameters being transformed/copied) because a user could pass a variety of parameters when creating the service - including params meant to customize the app deployment and params meant to customize a backing service deployment (e.g. `cf create-service ... -c '{"memory":"2G", "count":"3", "backingInstanceParam1":"value1", "backingInstanceParam2":"value2"}'`). You don't want all those params blindly applied to the backing service instance, so the broker developer should be explicit about which params are transformed by each parameters transformer.

**Comment by LittleBaiBai** [comment]
Follow up on Dylan's question, how will `include` handle nested properties? For params like:`{"git": {"uri": "some.uri", username: "some user"}}`, does `include: git` means everything under `git` get passed along, and `include: git.uri` only passes `"some.uri"`?

Can `include` do the jq style pattern matching and pass along all matching objects? For our use case we need `include: *`.

**Comment by AlbertoImpl** [comment]
@LittleBaiBai great question! By adding `include: git` everything under `git`, including `git`will be passed through.
I just created a branch with the current WIP. I demoed it to @pivotal-dylan-roberts a bit ago. There are many things I want to improve but the behavirour should not change, in case you want to test that out: https://github.com/spring-cloud-incubator/spring-cloud-app-broker/compare/139-backing-service-params
